config/internal/configsource/manager.go

```diff
@@ -295,74 +317,59 @@ func (m *Manager) retrieveEndAllSessions(ctx context.Context) []error {
+		// The value is of type []interface{} when an array is used in the configuration, YAML example:
+		//
+		//  array0:
+		//    - elem0
+		//    - elem1
+		//  array1:
+		//    - entry:
+		//        str: elem0
+		//	  - entry:
+		//        str: $tstcfgsrc:elem1
+		//
+		// Both "array0" and "array1" are going to be leaf config nodes hitting this case.
 		nslice := make([]interface{}, 0, len(v))
 		for _, vint := range v {
-			value, err := m.expandStringValues(ctx, vint)
+			value, err := m.parseConfigValue(ctx, vint)
 			if err != nil {
 				return nil, err
 			}
 			nslice = append(nslice, value)
 		}
 		return nslice, nil
 	case map[string]interface{}:
+		// The value is of type map[string]interface{} when an array in the configuration is populated with map
+		// elements. From the case above (for type []interface{}) each element of "array1" is going to hit the
+		// the current case block.
 		nmap := make(map[interface{}]interface{}, len(v))
 		for k, vint := range v {
-			value, err := m.expandStringValues(ctx, vint)
-			if err != nil {
-				return nil, err
-			}
-			nmap[k] = value
-		}
-		return nmap, nil
-	case map[interface{}]interface{}:
-		nmap := make(map[interface{}]interface{}, len(v))
-		for k, vint := range v {
-			value, err := m.expandStringValues(ctx, vint)
+			value, err := m.parseConfigValue(ctx, vint)
 			if err != nil {
 				return nil, err
 			}
 			nmap[k] = value
 		}
 		return nmap, nil
 	default:
+		// All other literals (int, boolean, etc) can't be further expanded so just return them as they are.
 		return v, nil
 	}
 }
 
-// expandConfigSource retrieve data from the specified config source and injects them into
-// the configuration. The Manager tracks sessions and watcher objects as needed.
-func (m *Manager) expandConfigSource(ctx context.Context, cfgSrc configsource.ConfigSource, s string) (interface{}, error) {
-	cfgSrcName, selector, params, err := parseCfgSrc(s)
-	if err != nil {
-		return nil, err
-	}
-
-	session, ok := m.sessions[cfgSrcName]
-	if !ok {
-		session, err = cfgSrc.NewSession(ctx)
-		if err != nil {
-			return nil, fmt.Errorf("failed to create session for config source %q: %w", cfgSrcName, err)
-		}
-		m.sessions[cfgSrcName] = session
-	}
-
-	retrieved, err := session.Retrieve(ctx, selector, params)
-	if err != nil {
-		return nil, fmt.Errorf("config source %q failed to retrieve value: %w", cfgSrcName, err)
-	}
-
-	m.watchers = append(m.watchers, retrieved.WatchForUpdate)
-
-	return retrieved.Value(), nil
-}
-
-// expandString expands environment variables and config sources that are specified on the string.
-func (m *Manager) expandString(ctx context.Context, s string) (interface{}, error) {
+// parseStringValue transforms environment variables and config sources, if any are present, on
+// the given string in the configuration into an object to be inserted into the resulting configuration.
+func (m *Manager) parseStringValue(ctx context.Context, s string) (interface{}, error) {
```

Comment:
Is it possible to split this func? It is long and with lots of nesting it is hard to read.

Reply:
In interest of time let's consider a refactor later: I can extract functions from some of the cases but they will make for very awkward parameters and function names. Added some comments to try to clarify them.